src/app/clusters/mode-select-server/mode-select-server.cpp

```diff
@@ -174,97 +257,230 @@ void emberAfModeSelectClusterServerInitCallback(EndpointId endpointId)
+
+bool Instance::HasFeature(ModeSelectFeature feature) const
 {
-    const EndpointId endpointId = attributePath.mEndpointId;
-    InteractionModel::Status result;
+    bool success;
+    uint32_t featureMap;
+    success = (GetFeature(&featureMap) == EMBER_ZCL_STATUS_SUCCESS);
 
-    switch (attributePath.mAttributeId)
+    return success && ((featureMap & to_underlying(feature)) != 0);
+}
+
+// List the commands supported by this instance.
+CHIP_ERROR Instance::EnumerateAcceptedCommands(const ConcreteClusterPath & cluster,
+                                               CommandHandlerInterface::CommandIdCallback callback, void * context)
+{
+    callback(ModeSelect::Commands::ChangeToMode::Id, context);
+    if (HasFeature(ModeSelect::ModeSelectFeature::kExt))
     {
-    case ModeSelect::Attributes::StartUpMode::Id:
-        result = verifyModeValue(endpointId, *value);
-        break;
-    case ModeSelect::Attributes::OnMode::Id:
-        result = verifyModeValue(endpointId, *value);
-        break;
-    default:
-        result = InteractionModel::Status::Success;
+        callback(ModeSelect::Commands::ChangeToModeWithStatus::Id, context);
+        callback(ModeSelect::Commands::ChangeToModeResponse::Id, context);
     }
 
-    return result;
+    return CHIP_NO_ERROR;
 }
 
-/**
- * Checks the new mode against the endpoint's supported modes.
- * @param endpointId    endpointId of the endpoint
- * @param newMode       value of the new mode
- * @return              Success status if the value is valid; InvalidValue otherwise.
- */
-static InteractionModel::Status verifyModeValue(const EndpointId endpointId, const uint8_t newMode)
+// Implements the read functionality for non-standard attributes.
+CHIP_ERROR Instance::Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder)
 {
-    if (NumericAttributeTraits<uint8_t>::IsNullValue(newMode)) // This indicates that the new mode is null.
+    switch (aPath.mAttributeId)
     {
-        return InteractionModel::Status::Success;
+    case Attributes::SupportedModes::Id:
+        if (msDelegate->modeOptions.empty())
+        {
+            aEncoder.EncodeEmptyList();
+            return CHIP_NO_ERROR;
+        }
+
+        CHIP_ERROR err;
+        const auto mo = msDelegate->modeOptions;
+        err           = aEncoder.EncodeList([mo](const auto & encoder) -> CHIP_ERROR {
+            for (ModeOptionStructType mode : mo)
+            {
+                ReturnErrorOnFailure(encoder.Encode(mode));
+            }
+            return CHIP_NO_ERROR;
+        });
+        ReturnErrorOnFailure(err);
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+// Implements checking before attribute writes.
+CHIP_ERROR Instance::Write(const ConcreteDataAttributePath & attributePath, AttributeValueDecoder & aDecoder)
+{
+    EmberAfStatus result;
+
+    if (aDecoder.WillDecodeNull()) // This indicates that the new mode is null.
```

Comment:
Please don't use WillDecodeNull().  Just decode into a Nullable<uint8_t>, which is the right type, and then see whether you got null.  WillDecodeNull() should really get removed, ideally; there's an issue tracking that.

Reply:
Fixed in [this commit](https://github.com/project-chip/connectedhomeip/pull/26447/commits/105c7bad401929e9a97a2ced738a003f6de70bc3).